app/Http/Controllers/ForgotPasswordController.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Carbon\Carbon;
+use Illuminate\Http\Request;
+use Illuminate\Support\Str;
+use Mail;
+
+use App\Models\User;
+use App\Models\PasswordResetToken;
+
+class ForgotPasswordController extends Controller
+{
+    public function __invoke() {
+        return view('auth.passwords.forget');
+    }
+
+    public function submitForgetPassword(Request $request) {
+        $request->validate([
+            'mail' => 'required|email',
+        ]);
+
+        // Do not tell the user that this email does not exist to prevent possible attacks
+        if (User::where('mail', $request->mail)->exists()) {
+            $token = Str::uuid();
+            PasswordResetToken::create([
+                'email' => $request->mail,
+                'token' => $token,
+                'created_at' => Carbon::now(),
+            ]);
+
+            Mail::send('email.forgetPassword', ['token' => $token], function($message) use($request) {
+                $message->to($request->mail);
+                $message->subject('Reset Password');
+            });
+        }
+
+        return back()->with('success', __('We have e-mailed your password reset link!'));
+    }
+
+    public function showResetPassword(Request $request, string $token) {
+        $prt = PasswordResetToken::select('token', 'created_at')->where('token', $token)->first();
+        if ($prt === null) {
+            return to_route('forgot.password.get')->withErrors('message', __('This token has expired!'));
+        }
+        $date = Carbon::now()->subMinutes(config('habbo.password_reset_token_time'));
+        if ($prt->created_at->gte($date)) { // gte = greater than or equals
+            $prt->delete();
+            return to_route('forgot.password.get')->withErrors('message', __('This token has expired!'));
+        }
+
+        return view('auth.passwords.reset', [
+            'token' => $token,
+        ]);
+    }
+
+    public function submitResetPassword(Request $request, string $token) {
+        $request->validate([
+            'password' => 'required|min:8|confirmed',
+            'password_confirmation' => 'required',
+        ]);
+
+        $prt = PasswordResetToken::select('email', 'token')->where('token', $token)->first();
+        if ($prt === null) {
+            return to_route('forgot.password.get')->withErrors('message', __('This token has expired!'));
+        }
+
+        $prt->user?->changePassword($request->password);
```

Comment:
```suggestion
        $prt->user->changePassword($request->password);
```

I don't think user can ever be null here?

Reply:
> The user can be deleted or the email can be changed by someone like the owner. If you change the password with the old email, you will receive an error message because that user no longer exists

if thats the case, then your system is flawed and should be revamped to not be able to change the email between a password request token lifetime of 15 minutes.